zarr/v3/codecs/__init__.py

```diff
@@ -110,48 +97,150 @@ def _validate_codecs(codecs: List[Codec], array_metadata: CoreArrayMetadata) ->
+            bb_codecs_with_spec,
+        ) = self._codecs_with_resolved_metadata(array_spec)
+
+        for bb_codec, array_spec in bb_codecs_with_spec[::-1]:
+            chunk_bytes = await bb_codec.decode(chunk_bytes, array_spec, runtime_configuration)
+
+        ab_codec, array_spec = ab_codec_with_spec
+        chunk_array = await ab_codec.decode(chunk_bytes, array_spec, runtime_configuration)
+
+        for aa_codec, array_spec in aa_codecs_with_spec[::-1]:
+            chunk_array = await aa_codec.decode(chunk_array, array_spec, runtime_configuration)
 
         return chunk_array
 
-    async def encode(self, chunk_array: np.ndarray) -> Optional[BytesLike]:
-        for aa_codec in self._array_array_codecs():
-            chunk_array_maybe = await aa_codec.encode(chunk_array)
+    async def decode_partial(
+        self,
+        store_path: StorePath,
+        selection: SliceSelection,
+        chunk_metadata: ArraySpec,
+        runtime_configuration: RuntimeConfiguration,
+    ) -> Optional[np.ndarray]:
+        assert self.supports_partial_decode
+        assert isinstance(self.array_bytes_codec, ArrayBytesCodecPartialDecodeMixin)
+        return await self.array_bytes_codec.decode_partial(
+            store_path, selection, chunk_metadata, runtime_configuration
+        )
+
+    async def encode(
+        self,
+        chunk_array: np.ndarray,
+        array_spec: ArraySpec,
+        runtime_configuration: RuntimeConfiguration,
+    ) -> Optional[BytesLike]:
+        (
+            aa_codecs_with_spec,
+            ab_codec_with_spec,
+            bb_codecs_with_spec,
+        ) = self._codecs_with_resolved_metadata(array_spec)
+
+        for aa_codec, array_spec in aa_codecs_with_spec:
+            chunk_array_maybe = await aa_codec.encode(
+                chunk_array, array_spec, runtime_configuration
+            )
             if chunk_array_maybe is None:
                 return None
             chunk_array = chunk_array_maybe
 
-        chunk_bytes_maybe = await self._array_bytes_codec().encode(chunk_array)
+        ab_codec, array_spec = ab_codec_with_spec
+        chunk_bytes_maybe = await ab_codec.encode(chunk_array, array_spec, runtime_configuration)
         if chunk_bytes_maybe is None:
             return None
         chunk_bytes = chunk_bytes_maybe
 
-        for bb_codec in self._bytes_bytes_codecs():
-            chunk_bytes_maybe = await bb_codec.encode(chunk_bytes)
+        for bb_codec, array_spec in bb_codecs_with_spec:
+            chunk_bytes_maybe = await bb_codec.encode(
+                chunk_bytes, array_spec, runtime_configuration
+            )
             if chunk_bytes_maybe is None:
                 return None
             chunk_bytes = chunk_bytes_maybe
 
         return chunk_bytes
 
-    def compute_encoded_size(self, byte_length: int) -> int:
-        return reduce(lambda acc, codec: codec.compute_encoded_size(acc), self.codecs, byte_length)
+    async def encode_partial(
```

Comment:
what would the consequences be for removing `encode`, renaming `encode_partial` to `encode`, and using using exceptions for codecs that can't do partial encoding (i.e., for these codecs, if `selection` is anything other than a full selection, raise an exception, which the caller can handle). logically, `encode_partial` and `decode_partial` are just more general versions of `encode` and `decode`, so it feels safe to rely on this fact in the API design.

Reply:
I think `encode` and `encode_partial` are different enough to keep them separate. `encode_partial` has a number of limitations, the main one being that it is currently only works with a single codec. Also, `encode_partial` takes a reference to the store to load the data itself; `encode` takes a buffer.